### PR TITLE
HARP-8604: Fix for multiple time tile loading using setTimeout() call.

### PR DESCRIPTION
**Pull request by kkostecki**
The decoded tiles geometry was loaded/generated multiple times, both with
techniques processing in SimpleTileGeometryLoader.
This patchset fixes the flow so the request for tile load after decoding is
processed only once.
It also adds some sanity checks that ensures us that disposed or unvisible
tiles are no longer loaded/processed.

Signed-off-by: Krystian Kostecki <ext-krystian.kostecki@here.com>

Thank you for contributing to harp.gl!

Before requesting a pull request, please remember to check the following documents:
* [contribution guidelines](https://github.com/heremaps/harp.gl/blob/master/CONTRIBUTING.md)
* [coding style](https://github.com/heremaps/harp.gl/blob/master/CODINGSTYLE.md)

If you are adding new functionality we would highly appreciate if you can describe what is the capability you are adding and even better if you can add some examples. Please also remember to add tests for it.

# CI Check

Our bots will check whether your PR can be directly integrated into the mainline. We have some internal integration tests running on the background, our bots will inform you of the next steps and someone from our team will take a look and help if needed!

And please do not forget to sign-off your commit! You can read more about DCO [here](https://julien.ponge.org/blog/developer-certificate-of-origin-versus-contributor-license-agreements/). But, in short, you just need to use `git commit -s` or append `--signoff` when you are committing to the repo.

Happy contributing!

### CODE REVIEW

File: @here/harp-mapview/lib/geometry/TileGeometryLoader.ts [hunk 1]
```diff
@@ -246,16 +256,33 @@ export class SimpleTileGeometryLoader implements TileGeometryLoader {
...
     ) {
+        if (this.isFinished) {
+            return;
+        }
         // If the tile is not ready for display, or if it has become invisible while being loaded,
         // for example by moving the camera, the tile is not finished and its geometry is not
         // created. This is an optimization for fast camera movements and zooms.
         const tile = this.tile;
         const decodedTile = this.m_decodedTile;
         this.m_decodedTile = undefined;
```
Comment (kkostecki) [inline @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: Anybody knows why this was set (garbage collector, rejections)? This is more or less the reason of entire problems, the isLoading method fixes it using different member, but ideally I would remove this line and change isLoading getter to:
```
get isLoading(): boolean {
return this.m_decodedTile !== undefined && this.m_timeout !== undefined;
}
```

Comment (nzjony) [reply @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: I also wondered about this, and even created a ticket for it: HARP-8742

Even though it is removed, it still exists in the TileLoader, so we aren't saving any memory.

File: @here/harp-mapview/lib/geometry/TileGeometryLoader.ts [hunk 2]
```diff
@@ -205,7 +209,7 @@ export class SimpleTileGeometryLoader implements TileGeometryLoader {
         const tile = this.tile;
 
         // First time this tile is handled:
-        if (this.m_decodedTile === undefined && tile.decodedTile !== undefined) {
+        if (this.m_decodedTile === undefined && tile.decodedTile !== undefined && !this.isLoading) {
```
Comment (kkostecki) [inline @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: Using ```!this.isLoading``` check here stops processing both techniques and tile geometry load, but this way we will need to _wait_ until timeout to cancel invisible or disposed tile loading. Other method would be to check this condition in the update section (as in prepareForRender) and if tile is no longer visible cancel timeout.
This may give us some small improvements so less timeouts may be queued while panning.

File: @here/harp-mapview/lib/geometry/TileGeometryLoader.ts [hunk 3]
```diff
@@ -229,10 +236,13 @@ export class SimpleTileGeometryLoader implements TileGeometryLoader {
         }
         if (this.m_timeout !== undefined) {
             clearTimeout(this.m_timeout);
+            this.m_timeout = undefined;
         }
     }
 
     private finish() {
+        // TODO: Would be easier to set m_decodedTile = undefined, then isLoading would check its
+        // value and if m_timeout is set.
```
Comment (kkostecki) [inline @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: In my opinion we should set:
```
this.m_decodedTile = undefined;
```
here.

File: @here/harp-mapview/lib/geometry/TileGeometryLoader.ts [hunk 4]
```diff
@@ -246,16 +256,33 @@ export class SimpleTileGeometryLoader implements TileGeometryLoader {
...
         const decodedTile = this.m_decodedTile;
         this.m_decodedTile = undefined;
+        // TODO: Entire block could be moved to update method
         if (decodedTile === undefined || tile.disposed || !tile.isVisible) {
+            // TODO: Should we clearTimer here?
+            if (this.m_timeout !== undefined) {
+                clearTimeout(this.m_timeout);
+            }
+            // TODO: Should we dispose tile here?
+            //if (!tile.isVisible) {
```
Comment (kkostecki) [inline @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: This is the block that could be moved to update() function in order to cancel loading (timeout) unnecessary tiles immediately, otherwise once loading started it will happen when timeout is processed.

File: @here/harp-mapview/lib/geometry/TileGeometryLoader.ts (hunk 1 above)
Comment (zbigg) [reply @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: I confirm that originally this was to save memory, so decoded tile is wiped after closure passed to SetTimeout disapears ... then `*GeometryLoaders` were born and everything changed.

File: @here/harp-mapview/lib/geometry/TileGeometryLoader.ts [hunk 5]
```diff
@@ -246,17 +286,28 @@ export class SimpleTileGeometryLoader implements TileGeometryLoader {
...
+            //    tile.mapView.visibleTileSet.disposeTile(tile);
+            //}
             this.finish();
             return;
         }
+
         this.m_timeout = setTimeout(() => {
+            this.m_timeout = undefined;
```
Comment (ninok) [inline @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: In my opinion we should not misuse the browser task queue as tile load queue.
This would all be much cleaner and more flexible if we would have a dedicated queue were we push new tiles to and only have the setTimeout fetching one tile from the queue.

In this way we would also be able to take the tile with the highest priority from the queue in the future.

File: @here/harp-mapview/lib/geometry/TileGeometryLoader.ts [hunk 6]
```diff
@@ -320,4 +378,73 @@ export class SimpleTileGeometryLoader implements TileGeometryLoader {
...
+     *
+     * Method compares input sets with recently used geometry kinds in performance wise
+     * manner, taking special care of undefined and zero size sets.
+     *
+     * @param enabledKinds Set of geometry kinds to be displayed or undefined.
+     * @param disabledKinds Set of geometry kinds that won't be rendered.
+     * @return `true` only if sets are logically equal, meaning that undefined and empty sets
+     * may result in same geometry (techniques kind) beeing rendered.
+     */
+    private compareGeometryKinds(
```
Comment (ninok) [inline @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: How is this related to the task queue.
I think nobody is using geometry kind filtering so I would rather drop that feature than adding even more code b/c of a feature nobody uses.

File: @here/harp-mapview/lib/geometry/TileGeometryLoader.ts (hunk 1 above)
Comment (kkostecki) [reply @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: So the newest implementation will make it undefined only after TileGeometryLoaded has finished his job. Because decoded tile it stored elsewhere, currently it does not give us any optimization.

File: @here/harp-mapview/lib/geometry/TileGeometryLoader.ts (hunk 6 above)
Comment (kkostecki) [reply @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: I would agree, but this was introduced for Phased Loading and actually never supported in SimpleLoader! This provides support for it.
In simple words, I have a bad habit that whenever I touch something I am trying to at least comply with the interfaces somebody have prepared.
If we want to remove this we will need to drop PhasedLoading too, I guess and as far as I know there is separate ticket for this.

File: @here/harp-mapview/lib/geometry/TileGeometryLoader.ts (hunk 5 above)
Comment (kkostecki) [reply @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: I fully agree, just keep in mind that this PR have been introduced to fix multiple buildings extrusion animation that we spot on Firefox..., just look where it goes. Seems that in buildings fix I will need to provide entire new architecture, both with dropping some old features (Phased Loading?). Do we really want to go this way?
Maybe separate tickets should be created for all that steps..

**Comment by kkostecki** [comment]
> I review when there is a test.

I will provide test, as soon as we agree on current architecture used...

File: @here/harp-mapview/lib/geometry/TileGeometryLoader.ts (hunk 6 above)
Comment (kkostecki) [reply @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: Just a side note, maybe instead of trying to support kinds update, we could simply leave it as it was but simply add:
```
assert(enabledKinds == undefined || enabledKinds.size() === 0);
assert(disabledKinds == undefined || disabledKinds.size() === 0);
```
but this would assume that no one is using it... I heard that this is not purely true...

File: @here/harp-mapview/lib/geometry/TileGeometryLoader.ts (hunk 5 above)
Comment (kkostecki) [reply @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: Separate ticket created for it.

File: @here/harp-mapview/lib/geometry/TileGeometryLoader.ts (hunk 6 above)
Comment (ninok) [reply @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: I like your second proposal to avoid adding more dead code.
Right now we anyways crash or assert when phased loading is enabled. So we should just ignore the `enablePhaseLoading` flag and never enable it.

File: @here/harp-mapview/lib/geometry/TileGeometryLoader.ts [hunk 7]
```diff
@@ -167,6 +169,13 @@ export class SimpleTileGeometryLoader implements TileGeometryLoader {
         return this.m_isFinished;
     }
 
+    get isLoading(): boolean {
```
Comment (ninok) [inline @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: Remark for refactoring later (i.e. proper TileLoadQueue)
- `isLoading` is confusing `geometryCreationPending` would be more clear
- We should have a proper state enum in the Tile like:
```
enum TileState {
    Loading, // request was send to the server
    Decoding, // decoding in the worker
    GeometryCreation, // THREE.js object creation
    Finished,
    Failed
}
```
Right now it is really hard to get the tile state b/c there are so many components that need to be checked (`TileLoader`, `TileGeometryLoader`, ...)

Comment (kkostecki) [reply @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: Ok done.
